File: src/translation/jsonToProlog.js
const { createOntologyHint, convertJsonToProlog } = require('./translationUtils');

async function jsonToProlog(naturalLanguageText, llmClient, model = 'gpt-3.5-turbo', ontologyTerms = [], feedback = null, returnFullResponse = false) {
  try {
    // REMOVED: if (!llmClient) return '';
    
    const ontologyHint = createOntologyHint(ontologyTerms);
    const feedbackHint = feedback ? `\n\nFeedback from previous attempt: ${feedback}\n\n` : '';

    const prompt = `Translate the following into JSON representation, then convert to Prolog.${ontologyHint}${feedbackHint}
Output ONLY valid JSON with: 
- "type" ("fact"/"rule"/"query")
- "head" with "predicate" and "args" array
- "body" array (for rules only) with elements having "predicate" and "args"

Examples:
{"type":"fact","head":{"predicate":"bird","args":["tweety"]}}
{"type":"rule","head":{"predicate":"has_wings","args":["X"]},"body":[{"predicate":"bird","args":["X"]}]}
{"type":"query","head":{"predicate":"can_migrate","args":["tweety"]}}

Input: ${naturalLanguageText}
Output:`;
    
    const response = await llmClient.chat.completions.create({
      model,
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.0,
      response_format: { type: "json_object" }
    });
    
    const jsonOutput = JSON.parse(response.choices[0].message.content.trim());
    const prologResult = convertJsonToProlog(jsonOutput);

    // NEW: Return full response object if requested for metrics tracking
    return returnFullResponse ? { ...response, choices: [{ ...response.choices[0], message: { content: prologResult } }] } : prologResult;
  } catch (error) {
    // Add raw output to error for better feedback in mcr.js translateWithRetry
    if (error.response && error.response.choices && error.response.choices[0] && error.response.choices[0].message) {
      error.rawOutput = error.response.choices[0].message.content.trim();
    }
    throw error;
  }
}

module.exports = jsonToProlog;

File: java/src/main/java/com/mcr/translation/JsonToProlog.java
package com.mcr.translation;

import com.google.gson.Gson;
import dev.langchain4j.model.chat.ChatLanguageModel;
import dev.langchain4j.model.output.Response;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class JsonToProlog implements TranslationStrategy {
    @Override
    public Map<String, Object> translate(String naturalLanguageText, Object llmClient, String model, List<String> ontologyTerms, String feedback) throws Exception {
        ChatLanguageModel chatModel = (ChatLanguageModel) llmClient;
        String prompt = "Translate the following into JSON representation, then convert to Prolog.\n\nAvailable ontology terms: " + String.join(", ", ontologyTerms) + "\n\n" +
                "Output ONLY valid JSON with:\n" +
                "- \"type\" (\"fact\"/\"rule\"/\"query\")\n" +
                "- \"head\" with \"predicate\" and \"args\" array\n" +
                "- \"body\" array (for rules only) with elements having \"predicate\" and \"args\"\n\n" +
                "Examples:\n" +
                "{\"type\":\"fact\",\"head\":{\"predicate\":\"bird\",\"args\":[\"tweety\"]}}\n" +
                "{\"type\":\"rule\",\"head\":{\"predicate\":\"has_wings\",\"args\":[\"X\"]},\"body\":[{\"predicate\":\"bird\",\"args\":[\"X\"]}]}\n" +
                "{\"type\":\"query\",\"head\":{\"predicate\":\"can_migrate\",\"args\":[\"tweety\"]}}\n\n" +
                "Input: " + naturalLanguageText + "\n" +
                "Output:";

        long startTime = System.currentTimeMillis();
        String jsonResponse = chatModel.generate(prompt);
        long endTime = System.currentTimeMillis();

        Gson gson = new Gson();
        Map<String, Object> jsonMap = gson.fromJson(jsonResponse, Map.class);
        String prolog = jsonToProlog(jsonMap);


        Map<String, Object> result = new HashMap<>();
        result.put("prolog", prolog);
        result.put("json", jsonResponse);
        result.put("promptTokens", 0L);
        result.put("completionTokens", 0L);
        result.put("latencyMs", endTime - startTime);

        return result;
    }

    private String jsonToProlog(Map<String, Object> jsonMap) {
        String type = (String) jsonMap.get("type");
        Map<String, Object> head = (Map<String, Object>) jsonMap.get("head");
        String headPredicate = (String) head.get("predicate");
        List<String> headArgs = (List<String>) head.get("args");
        String headArgsString = String.join(", ", headArgs);
        String prolog = headPredicate + "(" + headArgsString + ")";

        if ("rule".equals(type)) {
            List<Map<String, Object>> body = (List<Map<String, Object>>) jsonMap.get("body");
            String bodyString = body.stream().map(p -> {
                String pred = (String) p.get("predicate");
                List<String> args = (List<String>) p.get("args");
                return pred + "(" + String.join(", ", args) + ")";
            }).collect(Collectors.joining(", "));
            prolog += " :- " + bodyString;
        }

        if ("fact".equals(type) || "rule".equals(type)) {
            prolog += ".";
        }

        return prolog;
    }
}
